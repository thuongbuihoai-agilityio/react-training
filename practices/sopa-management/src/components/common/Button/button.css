.btn {
  cursor: pointer;
}

.btn-primary {
  padding: var(--px-normal) 0;
  width: 181px;
  background-color: var(--text-primary);
  border-radius:  var(--rounded-tertiary);
  border: none;
}

.btn-secondary,
.btn-tertiary {
  background-color: var(--dark);
  border-radius: var(--rounded-tertiary);
  color: var(--light);
  padding: var(--px-normal) 0;
  width: var(--w-medium);
  border: none;
}

.btn-tertiary {
  width: var(--w-large);
  padding: var(--px-normal) 0;
}

.btn-icon-primary {
  background-color: transparent;
  border: none;
}

.btn-icon-secondary {
  background-color: transparent;
  border-radius: 50%;
  border: 1px solid var(--border-primary);
  padding: 0 var(--px-base);
}

.btn-out-line-primary {
  width: var(--w-normal);
  padding: var(--p-extra-small) 0;
  border-radius: var(--rounded-secondary);
  color: var(--text-quaternary);
  border: 1px solid var(--text-quaternary);
  background: var(--light);
}

.btn-out-line-secondary {
  width: var(--w-extra-normal);
  padding: var(--px-base) 0;
  border-radius: var(--rounded-tertiary);
  border: 1px solid var(--light);
  color: var(--light);
  background: transparent;
  font-size: var(--text-extra-sm);
}
