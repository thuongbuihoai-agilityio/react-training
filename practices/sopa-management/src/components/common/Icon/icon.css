.icon-close,
.icon-facebook,
.icon-instagram,
.icon-twitter,
.icon-bag,
.icon-user,
.icon-esquire,
.icon-vogua,
.icon-human,
.icon-trash,
.icon-star {
  display: inline-block;
}


.icon-close,
.icon-facebook,
.icon-instagram,
.icon-twitter,
.icon-bag,
.icon-user {
  width: 24px;
  height: 24px;
}

.icon-close {
  background-image: url(../../../assets/icons/close.svg);
}

.icon-facebook {
  background-image: url(../../../assets/icons/facebook.svg);
}

.icon-instagram {
  background-image: url(../../../assets/icons/instagram.svg);
}

.icon-twitter {
  background-image: url(../../../assets/icons/twitter.svg);
}

.icon-bag {
  background-image: url(../../../assets/icons/shopping-bag.svg);
}

.icon-user {
  background-image: url(../../../assets/icons/user.svg);
}

.icon-esquire {
  width: 247px;
  height: 67px;
  background-image: url(../../../assets/icons/esquire.svg);
}

.icon-vogua {
  width: 224px;
  height: 58px;
  background-image: url(../../../assets/icons/vogua.svg);
}

.icon-human {
  width: 112px;
  height: 40px;
  background-image: url(../../../assets/icons/humans-of-new-york.svg);
}

.icon-trash {
  width: 20px;
  height: 20px;
  background-image: url(../../../assets/icons/trash.svg);
}

.icon-star {
  width: 16px;
  height: 16px;
  background-image: url(../../../assets/icons/star.svg);
}
