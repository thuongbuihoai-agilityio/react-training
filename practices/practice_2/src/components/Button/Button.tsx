import React, { memo } from "react";
import { ButtonProps } from "@/types/commonType/button";
import "./button.css";

const Button: React.FC<ButtonProps> = ({ text, className, onClick }) => {
  return (
    <button
      className={`btn btn__${className}`} onClick={onClick}>
      {text}
    </button>
  );
}

export default memo(Button);
