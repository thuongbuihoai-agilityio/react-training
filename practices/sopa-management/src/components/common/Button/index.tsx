import {
  ButtonHTMLAttributes,
  memo
} from 'react';

// Styles
import './button.css';

export enum ButtonType {
  default = '',
  primary = 'primary',
  secondary = 'secondary',
  tertiary = 'tertiary',
  btnIconPrimary = 'icon-primary',
  btnIconSecondary = 'icon-secondary',
  btnOutlinePrimary = 'out-line-primary',
  btnOutlineSecondary = 'out-line-secondary',
}

interface ButtonProps {
  ariaLabel?: string;
  children?: string | JSX.Element;
  type?: ButtonType;
  disable?: boolean;
  submit?: ButtonHTMLAttributes<HTMLButtonElement>['type'];
  className?: string;
  onClick?: () => void;
  onSubmit?: () => void;
}

const Button = ({
  ariaLabel = '',
  children = '',
  type = ButtonType.default,
  disable = false,
  submit,
  className = '',
  onClick,
  onSubmit
}: ButtonProps) => (
  <button
    data-testid='button'
    aria-label={ariaLabel}
    className={`${className} btn btn-${type}`}
    disabled={disable}
    type={submit}
    onClick={onClick}
    onSubmit={onSubmit}
  >
    {children}
  </button>
);

export default memo(Button);
