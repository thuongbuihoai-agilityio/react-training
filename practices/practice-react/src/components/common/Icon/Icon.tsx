import React, { memo, MouseEventHandler, useMemo } from "react";
import "./icon.css";

interface IconProps {
  iconName: string;
  onClick?: MouseEventHandler<HTMLElement>;
}

const Icon: React.FC<IconProps> = memo(({ iconName="filter", onClick }) => {
  const className = useMemo(() => {
    switch (iconName) {
      case "filter":
        return `icon-${iconName}`;
      case "plus":
        return `icon-${iconName}`;
      case "minus":
        return `icon-${iconName}`;
      case "cart":
        return `icon-${iconName}`;
      case "fb":
        return `icon-${iconName}`;
      case "twitter":
        return `icon-${iconName}`;
      case "instagram":
        return `icon-${iconName}`;
      default:
        return "icon"
    }
  }, [iconName]);

  return <i data-testid="icon" onClick={onClick} className={className} />;
});

export default Icon;
