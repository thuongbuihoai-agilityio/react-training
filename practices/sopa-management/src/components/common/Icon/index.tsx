import React, { memo, MouseEventHandler } from 'react';
import './icon.css';

export enum IconType {
  close = 'close',
  esquire = 'esquire',
  vogua = 'vogua',
  human = 'human',
  facebook = 'facebook',
  instagram = 'instagram',
  twitter = 'twitter',
  bag = 'bag',
  user = 'user',
  trash = 'trash',
  star = 'star'
}

interface IconProps {
  iconName?: IconType;
  className?: string;
  onClick?: MouseEventHandler<HTMLElement>;
}

const Icon: React.FC<IconProps> = ({
  iconName,
  className,
  onClick = () => {}
}) => (
  <div
    data-testid='icon'
    className={`${className} icon-${iconName}`}
    onClick={onClick}
  />
);

export default memo(Icon);