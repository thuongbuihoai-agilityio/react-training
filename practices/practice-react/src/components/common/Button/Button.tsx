import React, { memo, useMemo } from "react";
import Icon from "../Icon/Icon";
import "./button.css";

interface ButtonProps {
  type?: string;
  text: string;
  icon?: boolean;
  disabled?: boolean;
  onClick?: () => void;
}

const Button: React.FC<ButtonProps> = memo(({ text, type="primary", onClick, icon, disabled }) => {
  const className = useMemo(() => {
    switch (type) {
      case "primary":
        return ` btn-${type}`;
      case "light":
        return ` btn-${type}`;
      case "dark":
        return ` btn-${type}`;
      case "large":
        return ` btn-${type}`;
      case "outline-light":
        return ` btn-${type}`;
      case "outline-dark":
        return ` btn-${type}`;
      default:
        return "btn"
    }
  }, [type]);

  return (
    <button className={className} disabled={disabled} onClick={onClick}>
      {text}
      {icon && (
        <div className="btn btn-icon">
          <Icon iconName="filter" />
        </div>
      )}
    </button>
  );
});

export default Button;
