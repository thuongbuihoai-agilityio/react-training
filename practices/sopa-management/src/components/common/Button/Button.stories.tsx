import type { Meta, StoryObj } from '@storybook/react';

// Components
import Button, { ButtonType } from '.';
import Icon, { IconType } from '../Icon';

const meta: Meta<typeof Button> = {
  component: Button
};

export default meta;
type Story = StoryObj<typeof Button>;

export const ButtonPrimary: Story = {
  args: {
    children: 'Show more',
    type: ButtonType.primary
  }
};

export const ButtonSecondary: Story = {
  args: {
    children: 'Add to bag $15',
    type: ButtonType.secondary
  }
};

export const ButtonTertiary: Story = {
  args: {
    children: 'LOGIN',
    type: ButtonType.tertiary
  }
};

export const ButtonOutLinePrimary: Story = {
  args: {
    type: ButtonType.btnOutlinePrimary,
    children: 'Shoe Finder Quiz'
  }
};

export const ButtonOutLineSecondary: Story = {
  args: {
    type: ButtonType.btnOutlineSecondary,
    children: 'Enter email here for updates'
  }
};

export const ButtonIconPrimary: Story = {
  args: {
    type: ButtonType.btnIconPrimary,
    children: <Icon iconName={IconType.trash} />
  }
};

export const ButtonIconSecondary: Story = {
  args: {
    type: ButtonType.btnIconSecondary,
    children: <Icon iconName={IconType.trash} />
  }
};

export const ButtonDisable: Story = {
  args: {
    children: 'SHOW MORE',
    disable: true
  }
};
