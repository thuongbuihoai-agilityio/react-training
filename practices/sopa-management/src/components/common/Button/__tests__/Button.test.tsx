// Libs
import {
  render,
  screen
} from '@testing-library/react';
import '@testing-library/jest-dom';
import userEvent from '@testing-library/user-event';

// Components
import Button, { ButtonType } from '@common/Button';
import Icon, { IconType } from '@components/common/Icon';

describe('Button component', () => {
  test("Should render Button component with type 'default'", async () => {
    const { getByTestId } = render(<Button type={ButtonType.default} />);
    const button = getByTestId('button');
    expect(button).toBeInTheDocument();
  });

  test("Should render Button component with type 'disable'", async () => {
    render(<Button disable type={ButtonType.default} />);
    const clickBtn = screen.getByRole('button');
    expect(clickBtn).toBeInTheDocument();
  });

  test("Should render Button component with type 'primary'", async () => {
    render(<Button children='SHOW MORE' type={ButtonType.primary} />);

    const primaryBtn = screen.getByRole('button', { name: /SHOW MORE/i });
    expect(primaryBtn).toBeInTheDocument();
  });

  test("Should render Button component with type 'secondary'", async () => {
    render(<Button children='Add to bag' type={ButtonType.secondary} />);

    const secondaryBtn = screen.getByRole('button', { name: /Add to bag/i });
    expect(secondaryBtn).toBeInTheDocument();
  });

  test("Should render Button component with type 'tertiary'", async () => {
    render(<Button children='Add to bag' type={ButtonType.tertiary} />);

    const tertiaryBtn = screen.getByRole('button', { name: /Add to bag/i });
    expect(tertiaryBtn).toBeInTheDocument();
  });

  test("Should render Button component with type 'btnIconPrimary'", async () => {
    render(
      <Button
        children={<Icon iconName={IconType.trash} />}
        type={ButtonType.btnIconPrimary}
      />
    );

    const btnIconPrimary = screen.getByRole('button');
    expect(btnIconPrimary).toBeInTheDocument();
  });

  test("Should render Button component with type 'btnIconSecondary'", async () => {
    render(
      <Button
        children={<Icon iconName={IconType.trash} />}
        type={ButtonType.btnIconSecondary}
      />
    );

    const btnIconSecondary = screen.getByRole('button');
    expect(btnIconSecondary).toBeInTheDocument();
  });

  test('Should render Button component with default value', async () => {
    render(<Button />);

    const button = screen.getByRole('button');
    expect(button).toBeInTheDocument();
  });

  test('Should render Button component with action onClick', async () => {
    const onClickButton = jest.fn();
    render(<Button onClick={onClickButton} type={ButtonType.secondary} />);

    const actionButton = screen.getByRole('button');
    await userEvent.click(actionButton);
    expect(onClickButton).toHaveBeenCalled();
  });

  test('Matches snapshot', () => {
    const { container } = render(
      <Button children='SHOW MORE' type={ButtonType.default} />
    );
    expect(container).toMatchSnapshot();
  });
});
