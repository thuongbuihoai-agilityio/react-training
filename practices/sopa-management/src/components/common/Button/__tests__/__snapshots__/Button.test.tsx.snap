// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button component Matches snapshot 1`] = `
<div>
  <button
    aria-label=""
    class=" btn btn-"
    data-testid="button"
  >
    SHOW MORE
  </button>
</div>
`;
